src/canvas/renderCanvas.js

```diff
@@ -76,18 +79,126 @@ import {REACT_TOTAL_NUM_LANES} from '../constants';
   context.fillRect(0, 0, canvasWidth, canvasHeight);
 }
 
-/**
- * Render all charting data (once it's loaded and processed) within the
- * "scrollable" region.
- */
-// TODO (windowing) We can avoid rendering all of this if we've scrolled some of it off screen.
-function renderReactMarksAndMeasures(
+function renderReactEventRow(
+  context,
+  events,
+  state,
+  hoveredEvent,
+  canvasWidth,
+  canvasHeight,
+  canvasStartY,
+): number {
+  const {offsetY} = state;
+
+  // Draw events
+  const baseY = canvasStartY + REACT_EVENT_ROW_PADDING;
+  events.forEach(event => {
+    const showHoverHighlight = hoveredEvent && hoveredEvent.event === event;
+    renderSingleReactEvent(
+      context,
+      state,
+      event,
+      canvasWidth,
+      baseY,
+      offsetY,
+      showHoverHighlight,
+    );
+  });
+
+  // Draw the hovered and/or selected items on top so they stand out.
+  // This is helpful if there are multiple (overlapping) items close to each other.
+  if (hoveredEvent !== null && hoveredEvent.event !== null) {
+    renderSingleReactEvent(
+      context,
+      state,
+      hoveredEvent.event,
+      canvasWidth,
+      baseY,
+      offsetY,
+      true,
+    );
+  }
+
+  const rowHeight =
+    REACT_EVENT_ROW_PADDING + REACT_EVENT_SIZE + REACT_EVENT_ROW_PADDING;
+
+  // Render bottom border
+  context.fillStyle = COLORS.PRIORITY_BORDER;
+  context.fillRect(
+    0,
+    Math.floor(canvasStartY + rowHeight - offsetY - REACT_PRIORITY_BORDER_SIZE),
+    canvasWidth,
+    REACT_PRIORITY_BORDER_SIZE,
+  );
+
+  return canvasStartY + rowHeight;
+}
+
+function renderSingleReactEvent(
+  context,
+  state,
+  event,
+  canvasWidth,
+  baseY,
+  panOffsetY,
+  showHoverHighlight,
+) {
+  const {timestamp, type} = event;
+
+  const x = timestampToPosition(timestamp, state);
+  if (x + EVENT_SIZE / 2 < 0 || canvasWidth < x) {
+    return; // Not in view
+  }
+
+  let fillStyle = null;
+
+  switch (type) {
+    case 'schedule-render':
+    case 'schedule-state-update':
+    case 'schedule-force-update':
+      if (event.isCascading) {
+        fillStyle = showHoverHighlight
+          ? COLORS.REACT_SCHEDULE_CASCADING_HOVER
+          : COLORS.REACT_SCHEDULE_CASCADING;
+      } else {
+        fillStyle = showHoverHighlight
+          ? COLORS.REACT_SCHEDULE_HOVER
+          : COLORS.REACT_SCHEDULE;
+      }
+      break;
+    case 'suspense-suspend':
+    case 'suspense-resolved':
+    case 'suspense-rejected':
+      fillStyle = showHoverHighlight
+        ? COLORS.REACT_SUSPEND_HOVER
+        : COLORS.REACT_SUSPEND;
+      break;
+    default:
+      console.warn(`Unexpected event type "${type}"`);
+      break;
+  }
+
+  if (fillStyle !== null) {
+    const circumference = REACT_EVENT_SIZE;
+    const y = baseY + REACT_EVENT_SIZE / 2 - panOffsetY;
+
+    context.beginPath();
+    context.fillStyle = fillStyle;
+    context.arc(x, y, circumference / 2, 0, 2 * Math.PI);
+    context.fill();
+  }
+}
+
+function renderReactMeasures(
```

Comment:
Comment block explaining what measures are would help here

Reply:
This should also indicate `: number` for return type

Reply:
👍 I'll add JSDoc comments to the above functions in this PR, and I've added `: number` in the 1/n PR